benchmarks/pool_benchmarks.cpp

```diff
@@ -0,0 +1,526 @@
+ * or SMALL (about 1GB) in total for benchmark measurements.
+ */
+#define LARGE 12000000000
+//#define MEDIUM 6000000000
+//#define SMALL 1000000000
+
+class AllocatorBenchmark : public benchmark::Fixture {
+public:
+  using ::benchmark::Fixture::SetUp;
+  using ::benchmark::Fixture::TearDown;
+
+  virtual void* allocate(std::size_t nbytes) = 0;
+  virtual void deallocate(void* ptr) = 0;
+  virtual void release() = 0;
+
+  void allocation(benchmark::State& st) {
+    const std::size_t size{static_cast<std::size_t>(st.range(0))};
+    std::size_t i{0};
+
+    Max_Allocations = setBounds(size);
+    m_allocations.resize(Max_Allocations);
+
+    while (st.KeepRunning()) {
+      if (i == Max_Allocations) {
+        st.PauseTiming();
+        for (std::size_t j{0}; j < Max_Allocations; j++)
+          deallocate(m_allocations[j]);
+        i = 0;
+        st.ResumeTiming();
+      }
+      m_allocations[i++] = allocate(size);
+    }
+
+    st.counters["BytesProcessed"] = benchmark::Counter(st.range(0), 
+                                    benchmark::Counter::kIsIterationInvariantRate, benchmark::Counter::OneK::kIs1024);
+
+    for (std::size_t j{0}; j < i; j++)
+      deallocate(m_allocations[j]);
+
+    release();
+  }
+
+  void deallocation_same_order(benchmark::State& st) {
+    const std::size_t size{static_cast<std::size_t>(st.range(0))};
+    std::size_t i{0};
+
+    Max_Allocations = setBounds(size);
+    m_allocations.resize(Max_Allocations);
+
+    while (st.KeepRunning()) {
+      if (i == 0 || i == Max_Allocations) {
+        st.PauseTiming();
+        for (std::size_t j{0}; j < Max_Allocations; j++)
+          m_allocations[j] = allocate(size);
+        i = 0;
+        st.ResumeTiming();
+      }
+      deallocate(m_allocations[i++]);
+    }
+
+    st.counters["BytesProcessed"] = benchmark::Counter(st.range(0), 
+                                    benchmark::Counter::kIsIterationInvariantRate, benchmark::Counter::OneK::kIs1024);
+
+    for (std::size_t j{i}; j < Max_Allocations; j++)
+      deallocate(m_allocations[j]);
+    
+    release();
+  }
+
+  void deallocation_reverse_order(benchmark::State& st) {
+    const std::size_t size{static_cast<std::size_t>(st.range(0))};
+    
+    Max_Allocations = setBounds(size);
+    int i{int(Max_Allocations)};
+    m_allocations.resize(Max_Allocations);
+
+    while (st.KeepRunning()) {
+      if (i == 0 || i == int(Max_Allocations)) {
+        st.PauseTiming();
+        for (int j{0}; j < int(Max_Allocations); j++) 
+          m_allocations[j] = allocate(size);
+        i = Max_Allocations;
+        st.ResumeTiming();
+      }
+      deallocate(m_allocations[--i]);
+    }
+
+    st.counters["BytesProcessed"] = benchmark::Counter(st.range(0), 
+                                    benchmark::Counter::kIsIterationInvariantRate, benchmark::Counter::OneK::kIs1024);
+
+    //for (int j{i}; j < int(Max_Allocations); j++)
+    //  deallocate(m_allocations[j]);
```

Comment:
This part still needs work - it was messing a few things up before and I forgot to add it back in.. stay tuned

Reply:
Will add these statements back in!